devtools/projects/shell-browser/src/app/background.ts

```
@@ -36,152 +36,214 @@ browserAction.setIcon(
+        topic: 'contentScriptConnected',
+        args: [contentScriptConnection.frameId, contentScriptConnection.port!.name]
+      });
+    }
+
+    if (!contentScriptConnection.enabled) {
       return;
     }
 
-    // tslint:disable-next-line:no-console
-    console.log('Content script connected', port.sender.tab.id);
-    tab = port.sender.tab.id.toString();
-    name = 'content-script';
-  }
+    devtoolsPort.postMessage(message);
+  };
+  contentScriptPort.onMessage.addListener(onContentScriptMessage);
 
-  let portsTab = ports[tab];
-  if (!portsTab) {
-    // tslint:disable-next-line:no-console
-    console.log('Creating a tab port');
+  const shutdownContentScript = () => {
+    devtoolsPort.onMessage.removeListener(onDevToolsMessage);
+    devtoolsPort.postMessage({
+      topic: 'contentScriptDisconnected',
+      args: [contentScriptConnection.frameId, contentScriptConnection.port!.name]
+    });
+
+    contentScriptPort.onMessage.removeListener(onContentScriptMessage);
+    contentScriptPort.disconnect();
+  };
+  contentScriptPort.onDisconnect.addListener(() => shutdownContentScript());
 
-    portsTab = ports[tab] = {
-      devtools: null,
-      'content-script': null,
-    };
+  contentScriptConnection.pipeSetup = true;
+}
+
+function getPopUpName(ng: AngularDetection) {
+  if (!ng.isAngular) {
+    return 'not-angular.html';
   }
+  if (!ng.isIvy || !ng.isSupportedAngularVersion) {
+    return 'unsupported.html';
+  }
+  if (!ng.isDebugMode) {
+    return 'production.html';
+  }
+  return 'supported.html';
+}
 
-  portsTab[name] = port;
+chrome.runtime.onConnect.addListener((port) => {
+  if (isNumeric(port.name)) {
+    registerPortTab(port, 'devtools', port.name);
+    return;
+  }
 
-  if (portsTab.devtools && portsTab['content-script']) {
-    doublePipe(portsTab.devtools, portsTab['content-script'], tab);
+  if (!port.sender || !port.sender.tab || port.sender.tab.id === undefined ||
+      port.sender.frameId === undefined) {
+    return;
```

Question: Do we expect ports to sometimes not have `tabId` or `frameId`? Should we throw an error / log something in this case?

---

Good point, added a warning log for this case. In practice this case shouldn't happen